### PR TITLE
Layer data events before and after

### LINKED ISSUE #6177: Emit layer.events.data before and after
## 🧰 Task
At the moment, the timepoint at which `layer.events.data` emits is inconsistent in both `Shapes` and `Points`. The task is to emit data both before and after `layer.data` gets edited using the present participle and the past participle form of the actions in `ActionType`.

For reference also see this post on [forum.image.sc](https://forum.image.sc/t/shapes-layer-events-data-connect-seems-to-be-triggered-at-the-beginning-of-drawing-new-shape/85182/5)

### PR DESCRIPTION
**Pull request by melonora**
# References and relevant issues
Closes #6177 

# Description
This PR fixes the issue of `layer.events.data` not emitting in a consistent manner. For example in case of adding a shape, an emit would happen at the start of drawing a shape, while for removing data it would happen after data was removed. This PR fixes the issue by changing `ActionType` in present participle and past participle and emitting before and after `layer.data` has been edited.
Furthermore, the way `layer.events.data` emits in the `data.setter` of `Shapes` and `Points` has changed. The behaviour is as follows:

1. If `layer.data` is populated and new data is not empty use `ActionType.CHANGING` and `ActionType.CHANGED`. In this case the `data_indices` correspond to the indices of the original data.
2. In case `layer.data` is not populated and new data is not empty use `ActionType.ADDING` and `ActionType.ADDED`. The `data_indices` correspond to the data being added.
3. In case new data is empty use `ActionType.REMOVING` and `ActionType.REMOVED`. The `data_indices` correspond to the original data

# Final Checklist
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to docstrings and documentation
  (open a PR on the docs repository (https://github.com/napari/docs) if relevant!)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans._("some string")` to make them localizable.
  (For more information see our [translations guide](https://napari.org/developers/translations.html)).

### CODE REVIEW

**Comment by melonora** [comment]
@jni I think we also need to adjust the emitter in the `data.setter`. Thinking about it I think we need to do the following:

1. If `layer.data` is populated use `ActionType.CHANGING` and `ActionType.CHANGED`. In this case the `data_indices` correspond to the indices of the original data.
2. In case `layer.data` is not populated use ``ActionType.ADDING` and `ActionType.ADDED`. The `data_indices` correspond to the data being added.

What do you think?

File: napari/layers/points/_points_mouse_bindings.py [hunk 1]
```diff
@@ -68,6 +69,16 @@ def select(layer, event):
...
+            # only emit just before moving
+            if not is_moving:
+                layer.events.data(
+                    value=layer.data,
+                    action=ActionType.CHANGING,
+                    data_indices=tuple(
+                        layer.selected_data,
+                    ),
+                    vertex_indices=((),),
+                )
```
Comment (Czaki) [inline napari/layers/points/_points_mouse_bindings.py:81]: Part of the function that should be split on smaller parts in the future. This change looks correctly but is not easy to load whole scope

File: napari/layers/points/points.py [hunk 2]
```diff
@@ -536,13 +536,56 @@ def data(self) -> np.ndarray:
...
+                vertex_indices=((),),
+            )
+        else:
+            self.events.data(
+                value=self.data,
+                action=ActionType.REMOVING,
+                data_indices=tuple(i for i in range(len(self.data))),
+                vertex_indices=((),),
+            )
+        self._set_data(data)
```
Comment (Czaki) [inline napari/layers/points/points.py]:
```suggestion
        kwargs = {"value": self.data, "vertex_indices": ((),), "data_indices": tuple(i for i in range(len(self.data)))}
        if prior_data and data_not_empty:
            kwargs["action"] = ActionType.CHANGING
        elif data_not_empty:
            kwargs["action"] = ActionType.ADDING
            kwargs["data_indices"] = tuple(i for i in range(len(data)))
        else:
            kwargs["action"] = ActionType.REMOVING
        self.events.data(**kwargs)
        self._set_data(data)
```

File: napari/layers/points/points.py [hunk 3]
```diff
@@ -536,13 +536,56 @@ def data(self) -> np.ndarray:
...
+                data_indices=tuple(i for i in range(len(data))),
+                vertex_indices=((),),
+            )
+        else:
+            self.events.data(
+                value=self.data,
+                action=ActionType.REMOVED,
+                data_indices=tuple(i for i in range(len(self.data))),
+                vertex_indices=((),),
+            )
```
Comment (Czaki) [inline napari/layers/points/points.py]: This is try to make it more readable by allowing see whole logit at once:

```suggestion
        kwargs = {"value": self.data, "vertex_indices": ((),), "data_indices": tuple(i for i in range(len(self.data)))}
        if prior_data and data_not_empty:
            kwargs["action"] = ActionType.CHANGING
        elif data_not_empty:
            kwargs["action"] = ActionType.ADDING
            kwargs["data_indices"] = tuple(i for i in range(len(data)))
        else:
            kwargs["action"] = ActionType.REMOVING
        self.events.data(**kwargs)
        self._set_data(data)
        kwargs["data_indices"] = tuple(i for i in range(len(data)))
        kwargs["data"] = self.data

        if prior_data and data_not_empty:
            kwargs["action"] = ActionType.CHANGED
        elif data_not_empty:
            kwargs["action"] = ActionType.ADDED
        else:
            kwargs["action"] = ActionType.REMOVED
        self.events.data(**kwargs)

```

File: napari/layers/shapes/shapes.py [hunk 4]
```diff
@@ -668,7 +668,33 @@ def data(self, data):
...
+                data_indices=tuple(i for i in range(len(data))),
+                vertex_indices=((),),
+            )
+        else:
+            self.events.data(
+                value=self.data,
+                action=ActionType.REMOVING,
+                data_indices=tuple(i for i in range(len(self.data))),
+                vertex_indices=((),),
+            )
```
Comment (Czaki) [inline napari/layers/shapes/shapes.py]: Same as above? Some try to get shorter code?

File: napari/layers/shapes/shapes.py [hunk 5]
```diff
@@ -1997,12 +2045,21 @@ def add(
                 z_index=z_index,
                 n_new_shapes=n_new_shapes,
             )
-            self.events.data(
-                value=self.data,
-                action=ActionType.ADD.value,
-                data_indices=(-1,),
-                vertex_indices=((),),
-            )
+            # This should only emit when programmatically adding as with drawing this leads to premature emit.
```
Comment (Czaki) [inline napari/layers/shapes/shapes.py:2031]: But mode still could be set when someone interacts via API.

File: napari/layers/points/points.py (hunk 3 above)
Comment (melonora) [reply napari/layers/points/points.py]: Thanks yes I agree!

File: napari/layers/shapes/shapes.py (hunk 5 above)
Comment (melonora) [reply napari/layers/shapes/shapes.py:2031]: I adjusted it now by adding the argument `gui` which only is `True` when adding a shape through `_shapes_mouse_bindings`.

File: napari/layers/points/_points_mouse_bindings.py (hunk 1 above)
Comment (melonora) [reply napari/layers/points/_points_mouse_bindings.py:81]: Yeah in general the different layers would benefit from refactoring. Part of that is on the roadmap.

**Comment by melonora** [comment]
Once this PR is approved I will update the documentation.

**Review by Czaki — APPROVED** [review]
Lenght some of this functions make pain during review but it looks correct. 

We need to find a way to avoid so function to grow so much. Need to add some check to CI.

**Pull request merged**